Fix kubectl drain error handling bug.

brianpursley opened the pull request:
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
Fixed a bug where kubectl drain would consider a pod as having been deleted if an error occurs while calling the API.

This can happen if you reboot the master node(s) while draining, for example.  In this case an error occurs, but the pod has not actually been deleted yet.

The problem occurs because the client-go returns an empty struct instead of nil, even when there is an error, as shown here:
https://github.com/kubernetes/kubernetes/blob/6aac45ff1e99068e834ba3b93b673530cf62c007/staging/src/k8s.io/client-go/kubernetes/typed/core/v1/pod.go#L75-L85

The kubectl drain code assumed that an error did not occur if the returned pod was not nil, which is incorrect.

This PR fixes that bug and updates the existing unit tests to use an empty struct instead of nil, matching what is actually happening.  I also added a new test case to cover the case where a pod is returned, but the UID is different.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes https://github.com/kubernetes/kubectl/issues/1532

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
```release-note
Fixed a bug where kubectl drain would consider a pod as having been deleted if an error occurs while calling the API.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
```docs

```

brianpursley commented:
/hold for review

/cc @eddiezane 
/cc @ardaguclu

ardaguclu commented:
/triage accepted
/priority backlog

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/drain/drain.go:
```diff
@@ -417,7 +417,7 @@ func waitForDelete(params waitForDeleteParams) ([]corev1.Pod, error) {
 		for i, pod := range pods {
 			p, err := params.getPodFn(pod.Namespace, pod.Name)
-			if apierrors.IsNotFound(err) || (p != nil && p.ObjectMeta.UID != pod.ObjectMeta.UID) {
+			if apierrors.IsNotFound(err) || (err == nil && p != nil && p.ObjectMeta.UID != pod.ObjectMeta.UID) {
```
As we discussed this in bug scrub, changes look reasonable to me. Do you think would it make sense to add a comment above this line describing the behavior?

brianpursley replied:
Good idea. Comment added.

It's actually unfortunate that it has this unnecessary `getPodFn` field because it obscures that fact that this behavior is caused by client-go.  In theory, another caller of `waitForDelete` could pass a different implementation that doesn't behave the same way.

In reality though, `waitForDeleteParams` is unexported and not used anywhere else, so it won't be different.

I think that is what this TODO comment is implying.  
https://github.com/kubernetes/kubernetes/blob/4aecb151b8301a6fc69f63c08f640c7b2b4289fc/staging/src/k8s.io/kubectl/pkg/drain/drain.go#L249-L252

A good refactor would be to remove `getPodFn` altogether and just call client-go directly from inside this function, eliminating the unnecessary wrapper function.

ardaguclu replied:
I fully agree that getPodFn is redundant and we can directly call client-go.

brianpursley committed:
Fix kubectl drain error handling bug.

Fixed a bug where kubectl drain would consider a pod as having been deleted if an error occurs while calling the API.

ardaguclu commented:
/approve

brianpursley commented:
/unhold

brianpursley commented:
@ardaguclu can you lgtm if you think this is good to go?

ardaguclu commented:
Thanks for the fix
/lgtm
